File: ios/Conversions.swift
import GoogleMobileAds
import UIKit

/// Mirror of the JS-side `BannerAdAdaptiveKind` (see `src/BannerAdSize.ts`). A size that carries
/// one of these must be rebuilt through the matching SDK factory: every adaptive kind is a flag
/// on `GADAdSize.flags`, and `adSizeFor(cgSize:)` produces a *fixed custom* size (flags = 1)
/// instead, which tells Google's serving side "custom WxH" rather than "adaptive".
///
/// `GADAdSize.h` says outright: "Do not create a GADAdSize manually... Do not access any fields
/// directly", so the flag cannot simply be copied onto a size built from the numbers — the
/// factory is the only supported way to get it.
///
/// The orientation is part of the kind because each anchored factory returns a different height
/// for the same width; rebuilding an explicitly portrait size through the current-orientation
/// factory on a landscape device would silently retarget the request.
enum BannerAdSizeKind: String {
  case anchored
  case anchoredPortrait
  case anchoredLandscape
  case largeAnchored
  case largeAnchoredPortrait
  case largeAnchoredLandscape
  case inline
}

/// The single place the SDK's adaptive size factories are called. Both the module's size
/// functions and `BannerAd`'s reconstruction of a size that crossed the JS boundary go through
/// here, so the two can never disagree — and the deprecated anchored factories are deprecated in
/// exactly one place.
///
/// Most of these are documented as main-queue-only, so every caller is already on main.
///
/// `maxHeight` is only read for `.inline`; the anchored factories derive their own height.
func makeAdaptiveAdSize(kind: BannerAdSizeKind, width: Double, maxHeight: Double) -> AdSize {
  switch kind {
  case .anchored:
    return currentOrientationAnchoredAdaptiveBanner(width: width)
  case .anchoredPortrait:
    return portraitAnchoredAdaptiveBanner(width: width)
  case .anchoredLandscape:
    return landscapeAnchoredAdaptiveBanner(width: width)
  case .largeAnchored:
    return largeAnchoredAdaptiveBanner(width: width)
  case .largeAnchoredPortrait:
    return largePortraitAnchoredAdaptiveBanner(width: width)
  case .largeAnchoredLandscape:
    return largeLandscapeAnchoredAdaptiveBanner(width: width)
  case .inline:
    return inlineAdaptiveBanner(width: width, maxHeight: maxHeight)
  }
}

/// Rebuilds the ad size a `BannerAdSize` describes. Falls back to a fixed custom size when the
/// size carries no adaptive marker (the fixed constants: BANNER, MEDIUM_RECTANGLE, ...), which is
/// exactly what those want.
func makeAdSize(adaptiveKind: String?, width: Double, height: Double) -> AdSize {
  guard let adaptiveKind, let kind = BannerAdSizeKind(rawValue: adaptiveKind) else {
    return adSizeFor(cgSize: CGSize(width: width, height: height))
  }
  return makeAdaptiveAdSize(kind: kind, width: width, maxHeight: height)
}

/// Converts `AdValuePrecision` to the JS-side `PaidEventValue.precision` string.
func adValuePrecisionToString(_ precision: AdValuePrecision) -> String {
  switch precision {
  case .unknown:
    return "unknown"
  case .estimated:
    return "estimated"
  case .publisherProvided:
    return "publisherProvided"
  case .precise:
    return "precise"
  @unknown default:
    return "unknown"
  }
}

func responseInfoToDictionary(_ info: ResponseInfo?) -> [String: Any?]? {
  guard let info else { return nil }
  return [
    "responseId": info.responseIdentifier,
    "mediationAdapterClassName": info.loadedAdNetworkResponseInfo?.adNetworkClassName,
    "adSourceName": info.loadedAdNetworkResponseInfo?.adSourceName,
    // Note: `GADAdNetworkResponseInfo` has no public `description` property. This used to
    // fill in NSObject's default `.description` here (not guaranteed to be meaningful,
    // effectively garbage), which has been removed. Dropping the `description` field from
    // the TypeScript-side `AdapterResponse` type is pending on the coordinator's side.
    "adapterResponses": info.adNetworkInfoArray.map { network -> [String: Any?] in
      [
        "adapterClassName": network.adNetworkClassName,
        "latencyMillis": Int(network.latency * 1000),
        "adError": network.error.map { errorToDictionary($0) },
      ]
    },
  ]
}

func errorToDictionary(_ error: Error) -> [String: Any?] {
  let nsError = error as NSError
  var dict: [String: Any?] = [
    "code": nsError.code,
    "message": nsError.localizedDescription,
    "domain": nsError.domain,
  ]
  // The GAD prefix isn't dropped on the Swift side either (it's a bare constant without
  // NS_TYPED_ENUM/NS_SWIFT_NAME).
  if let info = nsError.userInfo[GADErrorUserInfoKeyResponseInfo] as? ResponseInfo {
    dict["responseInfo"] = responseInfoToDictionary(info)
  }
  return dict
}

File: android/src/main/java/expo/modules/googlemobileads/Conversions.kt
package expo.modules.googlemobileads

import com.google.android.libraries.ads.mobile.sdk.common.AdSourceResponseInfo
import com.google.android.libraries.ads.mobile.sdk.common.AdValue
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError
import com.google.android.libraries.ads.mobile.sdk.common.MediationAdError
import com.google.android.libraries.ads.mobile.sdk.common.PrecisionType
import com.google.android.libraries.ads.mobile.sdk.common.ResponseInfo

// 実機の ads-mobile-sdk 1.4.0 (AAR を javap で確認) には、Task 6 brief が仮定していた
// `com.google.android.libraries.ads.mobile.sdk.common.AdError` は存在しない。
// トップレベルのロードエラーは `LoadAdError`、ミディエーションアダプター単位のエラーは
// `MediationAdError` に分かれており、フィールド構成も iOS の NSError ベースの AdError とは異なる。

fun ResponseInfo?.toMap(): Map<String, Any?>? {
  if (this == null) return null
  return mapOf(
    "responseId" to responseId,
    // 実際のフィールド名は adapterClassName / loadedAdSourceResponseInfo / adSourceResponses。
    // brief が仮定していた loadedAdapterResponseInfo / adapterResponses という名前は存在しない。
    "mediationAdapterClassName" to loadedAdSourceResponseInfo?.adapterClassName,
    "adSourceName" to loadedAdSourceResponseInfo?.name,
    "adapterResponses" to adSourceResponses.map { it.toMap() }
  )
}

private fun AdSourceResponseInfo.toMap(): Map<String, Any?> = mapOf(
  "adapterClassName" to adapterClassName,
  // AdSourceResponseInfo.latencyMillis は既にミリ秒単位で公開されている
  // （iOS の TimeInterval のような秒単位ではないため *1000 の変換は不要）。
  "latencyMillis" to latencyMillis,
  // description は意図的に含めない: AdSourceResponseInfo に description フィールドは無く、
  // JS 側の AdapterResponse 型も description を持たない（コーディネーターの指示通り）。
  "adError" to adError?.toMap()
)

// アダプター単位のエラー。実際に code: Int, message: String, domain: String を持つのはこちら。
private fun MediationAdError.toMap(): Map<String, Any?> = mapOf(
  "code" to code,
  "message" to message,
  "domain" to domain
)

// トップレベルのロードエラー。`LoadAdError` には domain フィールドが存在しないため
// （iOS の NSError.domain に相当するものが無い）、SDK 自身が公開している定数
// `MediationAdError.GOOGLE_MOBILE_ADS_DOMAIN` を「Google Mobile Ads SDK 自身が返した
// エラーである」ことを表す値として流用する。
fun LoadAdError.toMap(): Map<String, Any?> = mapOf(
  // code は `LoadAdError.ErrorCode` という enum であり、brief が仮定していた素の Int ではない。
  // JS 側の AdError.code は number なので、enum の getValue() で数値に変換する。
  "code" to code.value,
  "message" to message,
  "domain" to MediationAdError.GOOGLE_MOBILE_ADS_DOMAIN,
  "responseInfo" to responseInfo.toMap()
)

/** `PrecisionType` を JS 側の `PaidEventValue.precision` 文字列に変換する。 */
private fun PrecisionType.toJsString(): String = when (this) {
  PrecisionType.UNKNOWN -> "unknown"
  PrecisionType.ESTIMATED -> "estimated"
  PrecisionType.PUBLISHER_PROVIDED -> "publisherProvided"
  PrecisionType.PRECISE -> "precise"
}

/**
 * `paid` イベントのペイロードに変換する。`AdValue.valueMicros` はマイクロ単位の Long
 * （例: 1 米ドル = 1_000_000）なので、JS 側の `PaidEventValue.value`（number, 通貨の実額）
 * に合わせて 1,000,000 で割る。iOS 側は GADAdValue.value が既に実額の NSDecimalNumber
 * なので変換不要だったが、Android は単位が異なる。
 */
fun AdValue.toPaidEventMap(): Map<String, Any?> = mapOf(
  "value" to valueMicros / 1_000_000.0,
  "currencyCode" to currencyCode,
  "precision" to precisionType.toJsString()
)
